### PR TITLE
implement key bindings feature

### PR DESCRIPTION
**Pull request by kyokenn**
libratbag back-end part - https://github.com/libratbag/libratbag/pull/1368

### CODE REVIEW

**Comment by staticssleever668** [comment]
Does this make the macro-setting dialog fallback to a single key if full-fledged macros are not supported?
I think instead there should be a new dialog for key actions specifically, with an entry in the action list, just like macro does.

**Comment by kyokenn** [comment]
>Does this make the macro-setting dialog fallback to a single key if full-fledged macros are not supported?

No, I'm just using the first key from the macro.

> I think instead there should be a new dialog for key actions specifically, with an entry in the action list, just like macro does.

I would like to try to reuse the macro dialog and just limit it to single key action.

File: piper/buttondialog.py [hunk 1]
```diff
@@ -294,8 +296,13 @@ def _on_macro_updated(self, macro, pspec):
...
         # A macro has been set; update accordingly.
-        self._action_type = RatbagdButton.ActionType.MACRO
-        self._mapping = macro
+        if RatbagdButton.ActionType.MACRO in self._button.action_types:  # macro is supported
+            self._action_type = RatbagdButton.ActionType.MACRO
+            self._mapping = macro
+        else:
+            self._action_type = RatbagdButton.ActionType.KEY
+            type_, value = macro.keys[0]
+            self._mapping = value
```
Comment (staticssleever668) [inline piper/buttondialog.py:305]: > I would like to try to reuse the macro dialog and just limit it to single key action.

That does make sense.

Well, then I think this code should use the `key` action type opportunistically, falling back to macro and giving an error message if needed.

Pseudo-code:
```
if SupportsKeyActionType() and macro.keys.len() == 1:
    // SetKeyActionType()
    return
SetMacroActionType()
```

**Comment by kyokenn** [comment]
Changed buttons page so it's now shows key name instead of key code.

**Comment by staticssleever668** [comment]
One problem: when the type of button action is `key`, in the dialog with the list of possible button actions in `Send keystroke` it says `None`.

**Comment by kyokenn** [comment]
> One problem: when the type of button action is `key`, in the dialog with the list of possible button actions in `Send keystroke` it says `None`.

Fixed

**Review by staticssleever668 — APPROVED** [review]

**Pull request merged**